Add IPC deep dive (#616)
bitwarden/contributing-docs: docs/architecture/deep-dives/ipc.md

# Inter-Process Communication (IPC)

Bitwarden uses IPC to allow communication between and within certain parts of our clients. The

> oldest use-case is the communication between the Bitwarden browser extension and the Bitwarden

Reviewer: ```suggestion
original use-case is the communication between the Bitwarden browser extension and the Bitwarden
```

> oldest use-case is the communication between the Bitwarden browser extension and the Bitwarden

Author: Rephrased, resolving!

desktop application to allow the extension to be unlocked using biometric authentication.

> Bitwarden now has a generic framework for IPC provided in the SDK. This framework is used to provide

Reviewer: ```suggestion
Bitwarden now has a generic framework for IPC provided in our [SDK](https://github.com/bitwarden/sdk-internal). This framework is used to provide
```

Suggesting adding a link to disambiguate from our .NET SDK or SM SDK.

# Inter-Process Communication (IPC)

Bitwarden uses IPC to allow communication between and within certain parts of our clients. Examples
of IPC within a client include communication to and from the background script/service worker in the
browser extension and the main and renderer processes in the desktop application. IPC is also used
to allow communication between different clients, such as the browser extension and the desktop

> application to enable features like unlocking the extension using biometric authentication.

Reviewer: ```suggestion
application, to enable features like unlocking the extension using biometric authentication.
```

Bitwarden now has a generic framework for IPC provided in our
[SDK](https://github.com/bitwarden/sdk-internal). This framework is used to provide a common
interface for IPC across all clients. The framework is designed to be cross-platform and can be used
in any client that needs to communicate with another process.

## Usage

Please refer to the code documentation for more information on how to use the IPC framework:

- [IpcClient in the SDK](https://sdk-api-docs.bitwarden.com/bitwarden_ipc/struct.IpcClient.html)
- [IpcService in the TypeScript clients](https://github.com/bitwarden/clients/blob/main/libs/common/src/platform/ipc/ipc.service.ts)

## Availability

When fully rolled out, an initialized `IpcClient` will be available in all TypeScript clients

> through the `IpcService`, and in the SDK through the `BitwardenClient`. At the time of writing, the

Author: Made the suggested change but I'll keep an eye out to see if it causes any confusing :)
